### ISSUE TITLE
Support printing values other than integers

### ISSUE THREAD

**Issue by jaseemabid** [jaseemabid/Olifant#22]
The entry function called from C expects an integer, even though there is no specific reason for it to be so. This is just the first thing that worked. 

https://github.com/jaseemabid/Olifant/blob/695047f5a858683392262c1736552233c3ef08fd/runtime/olifant.c#L4

This prevents olifant from returning any other kinds of data types (like boolean) as the final result of the program. This sucks. 

## Option 1
Return a type and value to C. This should be the easiest way to ensure C interop, efficiency and using the best of machine capabilities. 

## Option 2
Encode runtime information into the value itself like Python's PyObject or using the last 3 bits of a word aligned pointer like [Ocaml](https://realworldocaml.org/v1/en/html/memory-representation-of-values.html). This is kind of cool but seems unnecessary in a completely statically typed language. The paper [incremental approach to compiler construction](./) talks in depth about using a similar memory model for the scheme implementation. 

I'm not sure which way to go.

**Comment by jaseemabid** [comment]
CC: @balajirrao @nachivpn

**Comment by nachivpn** [comment]
Why do you even need a "final result of a program" now that you have FFIs to do IO? Typically the final result would simply be an exit code, right?

**Comment by jaseemabid** [comment]
See how the result is printed. https://github.com/jaseemabid/Olifant/blob/695047f5a858683392262c1736552233c3ef08fd/runtime/olifant.c#L19

**Comment by nachivpn** [comment]
Ok, I thought you had an FFI to print values from Olifant. Why not add that? Are you trying to keep it "purely" functional?

**Comment by jaseemabid** [comment]
Not really. As of now, I don't have enough machinery in the language to implement a `show`. This whole thing could be thrown away once the lang can do that.

**Comment by nachivpn** [comment]
Check this out:

https://github.com/nachivpn/javalette/blob/master/lib/runtime.ll
